File: src/main/java/com/meltwater/puppy/RabbitPuppy.java
package com.meltwater.puppy;

import com.meltwater.puppy.config.BindingData;
import com.meltwater.puppy.config.ExchangeData;
import com.meltwater.puppy.config.PermissionsData;
import com.meltwater.puppy.config.QueueData;
import com.meltwater.puppy.config.RabbitConfig;
import com.meltwater.puppy.config.UserData;
import com.meltwater.puppy.config.VHostData;
import com.meltwater.puppy.config.reader.RabbitConfigException;
import com.meltwater.puppy.rest.RabbitRestClient;
import com.meltwater.puppy.rest.RestClientException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static java.lang.String.format;

public class RabbitPuppy {

    private static final Logger log = LoggerFactory.getLogger(RabbitPuppy.class);

    private final Pattern atVHostPattern = Pattern.compile("([^@]+)@([^@]+)");

    private final RabbitRestClient client;

    public RabbitPuppy(String brokerAddress, String username, String password) {
        client = new RabbitRestClient(brokerAddress, username, password);
    }

    public RabbitPuppy(RabbitRestClient client) {
        this.client = client;
    }

    public boolean waitForBroker(int seconds) {
        long start = new Date().getTime();
        while (new Date().getTime() < start + seconds*1000) {
            if (client.ping()) {
                return true;
            } else {
                log.info("Broker not available, waiting...");
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException ignored) {
            }
        }
        return false;
    }

    /**
     * Apply configuration to RabbitMQ Broker
     *
     * @param config Configuration to apply
     * @throws RabbitPuppyException If errors or configuration mismatches are encountered
     */
    public void apply(RabbitConfig config) throws RabbitPuppyException {
        List<Throwable> errors = new ArrayList<>();

        if (config.getVhosts().size() > 0)
            createVHosts(config.getVhosts(), errors);

        if (config.getUsers().size() > 0)
            createUsers(config.getUsers(), errors);

        if (config.getPermissions().size() > 0)
            createPermissions(config.getPermissions(), errors);

        if (config.getExchanges().size() > 0)
            createExchanges(config, errors);

        if (config.getQueues().size() > 0)
            createQueues(config, errors);

        if (config.getBindings().size() > 0)
            createBindings(config, errors);

        if (errors.size() > 0) {
            throw new RabbitPuppyException("Encountered errors while applying configuration", errors);
        }
    }

    /**
     * Create vhosts based on configuration.
     *
     * @param vhosts Configured vhosts
     * @return List of errors encountered during creation
     */
    private void createVHosts(Map<String, VHostData> vhosts, final List<Throwable> errors) {
        try {
            final Map<String, VHostData> existing = client.getVirtualHosts();
            vhosts.entrySet().forEach(entry -> {
                String name = entry.getKey();
                VHostData data = entry.getValue() == null ? new VHostData() : entry.getValue();
                ensurePresent("vhost", name, data, existing, errors, () -> {
                    log.info("Creating vhost " + entry.getKey());
                    client.createVirtualHost(entry.getKey(), data);
                });
            });
        } catch (RestClientException e) {
            log.error("Failed to fetch vhosts", e);
            errors.add(e);
        }
    }

    /**
     * Create users based on configuration.
     *
     * @param users Configured users
     * @return List of errors encountered during creation
     */
    private void createUsers(Map<String, UserData> users, final List<Throwable> errors) {
        try {
            final Map<String, UserData> existing = withKnownPasswords(client.getUsers(), users);
            users.entrySet().forEach(entry -> {
                String name = entry.getKey();
                UserData data = entry.getValue() == null ? new UserData() : entry.getValue();
                ensurePresent("user", name, data, existing, errors, () -> {
                    log.info("Creating user " + entry.getKey());
                    client.createUser(entry.getKey(), data);
                });
            });
        } catch (RestClientException e) {
            log.error("Failed to fetch vhosts", e);
            errors.add(e);
        }
    }

    /**
     * Creates user permissions per vhost based on configuration.
     *
     * @param permissions Configured permissions
     * @return List of errors encountered during creation
     */
    private void createPermissions(Map<String, PermissionsData> permissions, final List<Throwable> errors) {
        try {
            final Map<String, PermissionsData> existing = client.getPermissions();
            permissions.entrySet().forEach(entry -> {
                String name = entry.getKey();
                final Matcher matcher = atVHostPattern.matcher(name);
                if (matcher.matches()) {
                    PermissionsData data = entry.getValue() == null ? new PermissionsData() : entry.getValue();
                    ensurePresent("permissions", name, data, existing, errors, () -> {
                        String user = matcher.group(1);
                        String vhost = matcher.group(2);
                        log.info(format("Setting permissions for user %s at vhost %s", user, vhost));
                        client.createPermissions(user, vhost, data);
                    });
                } else {
                    String error = format("Invalid exchange format '%s', should be exchange@vhost", name);
                    log.error(error);
                    errors.add(new RabbitConfigException(error));
                }
            });
        } catch (RestClientException e) {
            log.error("Failed to fetch exchanges", e);
            errors.add(e);
        }
    }

    /**
     * Creates exchanges based on configuration.
     *
     * @param config Rabbit configuration
     * @return List of errors encountered during creation
     */
    private void createExchanges(RabbitConfig config, final List<Throwable> errors) {
        config.getExchanges().entrySet().forEach(entry -> {
            String name = entry.getKey();
            final Matcher matcher = atVHostPattern.matcher(name);
            if (matcher.matches()) {
                final String exchange = matcher.group(1);
                final String vhost = matcher.group(2);
                ExchangeData data = entry.getValue() == null ? new ExchangeData() : entry.getValue();
                log.debug(format("Ensuring exchange %s exists at vhost %s with configuration %s", exchange, vhost, data));
                final Auth auth = authForResource(config.getUsers(), config.getPermissions(), vhost, exchange);
                try {
                    Optional<ExchangeData> existing = client.getExchange(vhost, exchange, auth.getUser(), auth.getPass());
                    ensurePresent("exchange", name, data, existing, errors, () -> {
                        log.info(format("Creating exchange %s at vhost %s with configuration %s", exchange, vhost, data));
                        client.createExchange(vhost, exchange, data, auth.getUser(), auth.getPass());
                    });
                } catch (RestClientException e) {
                    log.error(format("Exception when ensuring exchange %s", exchange), e);
                    errors.add(e);
                }
            } else {
                String error = format("Invalid exchange format '%s', should be exchange@vhost", name);
                log.error(error);
                errors.add(new RabbitConfigException(error));
            }
        });
    }

    /**
     * Creates queues based on configuration.
     *
     * @param config Rabbit configuration
     * @return List of errors encountered during creation
     */
    private void createQueues(RabbitConfig config, final List<Throwable> errors) {
        config.getQueues().entrySet().forEach(entry -> {
            String name = entry.getKey();
            final Matcher matcher = atVHostPattern.matcher(name);
            if (matcher.matches()) {
                final String queue = matcher.group(1);
                final String vhost = matcher.group(2);
                QueueData data = entry.getValue() == null ? new QueueData() : entry.getValue();
                log.debug(format("Ensuring queue %s exists at vhost %s with configuration %s", queue, vhost, data));
                final Auth auth = authForResource(config.getUsers(), config.getPermissions(), vhost, queue);
                try {
                    Optional<QueueData> existing = client.getQueue(vhost, queue, auth.getUser(), auth.getPass());
                    ensurePresent("queue", name, data, existing, errors, () -> {
                        log.info(format("Creating queue %s at vhost %s with configuration %s", queue, vhost, data));
                        client.createQueue(vhost, queue, data, auth.getUser(), auth.getPass());
                    });
                } catch (RestClientException e) {
                    log.error(format("Exception when ensuring queue %s", queue), e);
                    errors.add(e);
                }
            } else {
                String error = format("Invalid queue format '%s', should be queue@vhost", name);
                log.error(error);
                errors.add(new RabbitConfigException(error));
            }
        });
    }

    private void createBindings(RabbitConfig config, final List<Throwable> errors) {
        config.getBindings().entrySet().forEach(entry -> {
            final String name = entry.getKey();
            final Matcher matcher = atVHostPattern.matcher(name);
            if (matcher.matches()) {
                final String exchange = matcher.group(1);
                final String vhost = matcher.group(2);
                final List<BindingData> bindings = (List<BindingData>) entry.getValue();
                final Auth auth = authForResource(config.getUsers(), config.getPermissions(), vhost, exchange);
                try {
                    final Map<String, List<BindingData>> existingVhost = client.getBindings(vhost, auth.user, auth.pass);
                    final List<BindingData> existing = existingVhost.containsKey(exchange) ?
                            existingVhost.get(exchange) :
                            new ArrayList<>();
                    bindings.forEach(bindingData -> {
                        log.info(format("Ensuring binding %s %s", name, bindingData));
                        if (!existing.contains(bindingData)) {
                            log.info(format("Creating binding %s : %s", name, bindingData));
                            try {
                                client.createBinding(vhost,
                                                     exchange,
                                                     bindingData,
                                                     auth.user,
                                                     auth.pass);
                            } catch (RestClientException e) {
                                log.error(format("Failed to create binding %s %s: %s", name, bindingData, e.getMessage()));
                                errors.add(e);
                            }
                        }
                    });
                } catch (RestClientException e) {
                    log.error(format("Failed to fetch bindings from %s", name), e);
                    errors.add(e);
                }
            } else {
                String error = format("Invalid binding format '%s', should be exchange@vhost", name);
                log.error(error);
                errors.add(new RabbitConfigException(error));
            }
        });
    }

    /**
     * Ensures that the configured resource is present on the broker.
     * Throws exception if creation failed, or resource exists with settings that does not match expected configuration.
     */
    private <D> void ensurePresent(String type, String name, D data, Map<String, D> existing, List<Throwable> errors, Create create) {
        if (existing.containsKey(name)) {
            if (!existing.get(name).equals(data)) {
                String error = format("%s '%s' exists but with wrong configuration: %s, expected: %s",
                        type, name, existing.get(name), data);
                log.error(error);
                errors.add(new InvalidConfigurationException(error));
            }
        } else {
            try {
                create.create();
            } catch (RestClientException e) {
                log.error(format("Failed to create %s '%s': %s", type, name, e.getMessage()));
                errors.add(e);
            }
        }
    }

    /**
     * Ensures that the configured resource is present on the broker.
     * Throws exception if creation failed, or resource exists with settings that does not match expected configuration.
     */
    private <D> void ensurePresent(String type, String name, D data, Optional<D> existing, List<Throwable> errors, Create create) {
        if (existing.isPresent()) {
            if (!existing.get().equals(data)) {
                String error = format("%s '%s' exists but with wrong configuration: %s, expected: %s",
                        type, name, existing.get(), data);
                log.error(error);
                errors.add(new InvalidConfigurationException(error));
            }
        } else {
            try {
                create.create();
            } catch (RestClientException e) {
                log.error(format("Failed to create %s '%s': %s", type, name, e.getMessage()));
                errors.add(e);
            }
        }
    }

    /**
     * Because lambdas.
     */
    private interface Create {
        void create() throws RestClientException;
    }

    /**
     * Copies known passwords onto users received from broker, since we get only password hash from it, so that
     * ensurePresent does not fail due to differences in the password field.
     *
     * @param existing   Existing users
     * @param fromConfig Users from input configuration
     * @return Existing users with known passwords appended
     */
    private Map<String, UserData> withKnownPasswords(Map<String, UserData> existing,
                                                     Map<String, UserData> fromConfig) {
        existing.forEach((user, data) -> {
            if (fromConfig.containsKey(user)) {
                data.setPassword(fromConfig.get(user).getPassword());
            }
        });
        return existing;
    }

    /**
     * Attempts to find a user in the given configuration with rights to configure the requested resource.
     *
     * @param permissions  user permissions
     * @param vhost        vhost name
     * @param resourceName resource name
     * @return Optional of user with creation rights, or Optional.empty() if not found.
     */
    private Auth authForResource(Map<String, UserData> users,
                                 Map<String, PermissionsData> permissions,
                                 String vhost,
                                 String resourceName) {
        Optional<String> found = permissions.entrySet().stream()
                .filter(entry -> {
                    final Matcher matcher = atVHostPattern.matcher(entry.getKey());
                    if (matcher.matches() && matcher.group(2).equals(vhost)) {
                        if (Pattern.compile(entry.getValue().getConfigure()).matcher(resourceName).matches()) {
                            return true;
                        }
                    }
                    return false;
                })
                .map(entry -> {
                    Matcher matcher = atVHostPattern.matcher(entry.getKey());
                    matcher.find(); // Causes matcher to actually parse regex groups
                    return matcher.group(1);
                })
                .findFirst();

        if (found.isPresent()) {
            String user = found.get();
            String pass = users.get(user).getPassword();
            return new Auth(user, pass);
        } else {
            return defaultAuth();
        }
    }

    private Auth defaultAuth() {
        return new Auth(client.getUsername(), client.getPassword());
    }

    private class Auth {
        private final String user;
        private final String pass;

        public Auth(String user, String pass) {
            this.user = user;
            this.pass = pass;
        }

        public String getUser() {
            return user;
        }

        public String getPass() {
            return pass;
        }
    }
}

File: src/main/java/com/meltwater/puppy/RabbitPuppy.kt
package com.meltwater.puppy

import com.meltwater.puppy.config.*
import com.meltwater.puppy.config.reader.RabbitConfigException
import com.meltwater.puppy.rest.RabbitRestClient
import com.meltwater.puppy.rest.RestClientException
import org.slf4j.LoggerFactory
import java.util.*
import java.util.regex.Pattern

class RabbitPuppy {

    private val log = LoggerFactory.getLogger(RabbitPuppy::class.java)

    private val atVHostPattern = Pattern.compile("([^@]+)@([^@]+)")

    private val client: RabbitRestClient

    constructor(brokerAddress: String, username: String, password: String) {
        client = RabbitRestClient(brokerAddress, username, password)
    }

    constructor(client: RabbitRestClient) {
        this.client = client
    }

    fun waitForBroker(seconds: Int): Boolean {
        val start = Date().time
        while (Date().time < start + seconds * 1000) {
            if (client.ping()) {
                return true
            } else {
                log.info("Broker not available, waiting...")
            }
            try {
                Thread.sleep(1000)
            } catch (ignored: InterruptedException) {
            }

        }
        return false
    }

    /**
     * Apply configuration to RabbitMQ Broker

     * @param config Configuration to apply
     * *
     * @throws RabbitPuppyException If errors or configuration mismatches are encountered
     */
    @Throws(RabbitPuppyException::class)
    fun apply(config: RabbitConfig) {
        val errors = ArrayList<Throwable>()

        if (config.vhosts.size > 0)
            createVHosts(config.vhosts, errors)

        if (config.users.size > 0)
            createUsers(config.users, errors)

        if (config.permissions.size > 0)
            createPermissions(config.permissions, errors)

        if (config.exchanges.size > 0)
            createExchanges(config, errors)

        if (config.queues.size > 0)
            createQueues(config, errors)

        if (config.bindings.size > 0)
            createBindings(config, errors)

        if (errors.size > 0) {
            throw RabbitPuppyException("Encountered errors while applying configuration", errors)
        }
    }

    /**
     * Create vhosts based on configuration.

     * @param vhosts Configured vhosts
     * *
     * @return List of errors encountered during creation
     */
    private fun createVHosts(vhosts: Map<String, VHostData>, errors: MutableList<Throwable>) {
        try {
            val existing: Map<String, VHostData> = client.getVirtualHosts()
            vhosts.entries.forEach { entry ->
                val name: String = entry.key
                val data: VHostData = entry.value
                ensurePresent("vhost", name, data, existing, errors, {
                    log.info("Creating vhost $name")
                    client.createVirtualHost(name, data)
                })
            }
        } catch (e: RestClientException) {
            log.error("Failed to fetch vhosts", e)
            errors.add(e)
        }

    }

    /**
     * Create users based on configuration.

     * @param users Configured users
     * *
     * @return List of errors encountered during creation
     */
    private fun createUsers(users: Map<String, UserData>, errors: MutableList<Throwable>) {
        try {
            val existing = withKnownPasswords(client.getUsers(), users)
            users.entries.forEach { entry ->
                val name = entry.key
                val data = entry.value
                ensurePresent("user", name, data, existing, errors, {
                    log.info("Creating user $name")
                    client.createUser(entry.key, data)
                })
            }
        } catch (e: RestClientException) {
            log.error("Failed to fetch vhosts", e)
            errors.add(e)
        }

    }

    /**
     * Creates user permissions per vhost based on configuration.

     * @param permissions Configured permissions
     * *
     * @return List of errors encountered during creation
     */
    private fun createPermissions(permissions: Map<String, PermissionsData>, errors: MutableList<Throwable>) {
        try {
            val existing = client.getPermissions()
            permissions.entries.forEach { entry ->
                val name = entry.key
                val matcher = atVHostPattern.matcher(name)
                if (matcher.matches()) {
                    val data = entry.value
                    ensurePresent<PermissionsData>("permissions", name, data, existing, errors, {
                        val user = matcher.group(1)
                        val vhost = matcher.group(2)
                        log.info("Setting permissions for user $user at vhost $vhost")
                        client.createPermissions(user, vhost, data)
                    })
                } else {
                    val error = "Invalid exchange format '$name', should be exchange@vhost"
                    log.error(error)
                    errors.add(RabbitConfigException(error))
                }
            }
        } catch (e: RestClientException) {
            log.error("Failed to fetch exchanges", e)
            errors.add(e)
        }

    }

    /**
     * Creates exchanges based on configuration.

     * @param config Rabbit configuration
     * *
     * @return List of errors encountered during creation
     */
    private fun createExchanges(config: RabbitConfig, errors: MutableList<Throwable>) {
        config.exchanges.entries.forEach { entry ->
            val name = entry.key
            val matcher = atVHostPattern.matcher(name)
            if (matcher.matches()) {
                val exchange = matcher.group(1)
                val vhost = matcher.group(2)
                val data = entry.value
                log.debug("Ensuring exchange $exchange exists at vhost $vhost with configuration $data")
                val auth = authForResource(config.users, config.permissions, vhost, exchange)
                try {
                    val existing = client.getExchange(vhost, exchange, auth.first, auth.second)
                    ensurePresent<ExchangeData>("exchange", name, data, existing, errors, {
                        log.info("Creating exchange $exchange at vhost $vhost with configuration $data")
                        client.createExchange(vhost, exchange, data, auth.first, auth.second)
                    })
                } catch (e: RestClientException) {
                    log.error("Exception when ensuring exchange $exchange", e)
                    errors.add(e)
                }

            } else {
                val error = "Invalid exchange format '$name', should be exchange@vhost"
                log.error(error)
                errors.add(RabbitConfigException(error))
            }
        }
    }

    /**
     * Creates queues based on configuration.

     * @param config Rabbit configuration
     * *
     * @return List of errors encountered during creation
     */
    private fun createQueues(config: RabbitConfig, errors: MutableList<Throwable>) {
        config.queues.entries.forEach { entry ->
            val name = entry.key
            val matcher = atVHostPattern.matcher(name)
            if (matcher.matches()) {
                val queue = matcher.group(1)
                val vhost = matcher.group(2)
                val data = entry.value
                log.debug("Ensuring queue $queue exists at vhost $vhost with configuration $data")
                val auth = authForResource(config.users, config.permissions, vhost, queue)
                try {
                    val existing = client.getQueue(vhost, queue, auth.first, auth.second)
                    ensurePresent<QueueData>("queue", name, data, existing, errors, {
                        log.info("Creating queue $queue at vhost $vhost with configuration $data")
                        client.createQueue(vhost, queue, data, auth.first, auth.second)
                    })
                } catch (e: RestClientException) {
                    log.error("Exception when ensuring queue $queue", e)
                    errors.add(e)
                }

            } else {
                val error = "Invalid queue format '$name', should be queue@vhost"
                log.error(error)
                errors.add(RabbitConfigException(error))
            }
        }
    }

    private fun createBindings(config: RabbitConfig, errors: MutableList<Throwable>) {
        config.bindings.entries.forEach { entry ->
            val name = entry.key
            val matcher = atVHostPattern.matcher(name)
            if (matcher.matches()) {
                val exchange = matcher.group(1)
                val vhost = matcher.group(2)
                val auth = authForResource(config.users, config.permissions, vhost, exchange)
                try {
                    val existingVhost = client.getBindings(vhost, auth.first, auth.second)
                    val existing = existingVhost.getOrElse(exchange, {ArrayList<BindingData>()})
                    entry.value.forEach { bindingData ->
                        log.info("Ensuring binding $name $bindingData")
                        if (!existing.contains(bindingData)) {
                            log.info("Creating binding $name : $bindingData")
                            try {
                                client.createBinding(vhost, exchange, bindingData, auth.first, auth.second)
                            } catch (e: RestClientException) {
                                log.error("Failed to create binding $name $bindingData: ${e.message}")
                                errors.add(e)
                            }

                        }
                    }
                } catch (e: RestClientException) {
                    log.error("Failed to fetch bindings from $name", e)
                    errors.add(e)
                }

            } else {
                val error = "Invalid binding format '$name', should be exchange@vhost"
                log.error(error)
                errors.add(RabbitConfigException(error))
            }
        }
    }

    /**
     * Ensures that the configured resource is present on the broker.
     * Throws exception if creation failed, or resource exists with settings that does not match expected configuration.
     */
    private fun <D> ensurePresent(type: String, name: String, data: D, existing: Map<String, D>, errors: MutableList<Throwable>, create: () -> Unit) {
        if (existing.containsKey(name)) {
            if (existing[name] != data) {
                val error = "$type '$name' exists but with wrong configuration: $existing, expected: $data"
                log.error(error)
                errors.add(InvalidConfigurationException(error))
            }
        } else {
            try {
                create()
            } catch (e: RestClientException) {
                log.error("Failed to create $type '$name': ${e.message}".format(type, name, e.message))
                errors.add(e)
            }

        }
    }

    /**
     * Ensures that the configured resource is present on the broker.
     * Throws exception if creation failed, or resource exists with settings that does not match expected configuration.
     */
    private fun <D> ensurePresent(type: String, name: String, data: D, existing: Optional<D>, errors: MutableList<Throwable>, create: () -> Unit) {
        if (existing.isPresent) {
            if (existing.get() != data) {
                val error = "$type '$name' exists but with wrong configuration: ${existing.get()}, expected: $data"
                log.error(error)
                errors.add(InvalidConfigurationException(error))
            }
        } else {
            try {
                create()
            } catch (e: RestClientException) {
                log.error("Failed to create $type '$name': ${e.message}")
                errors.add(e)
            }

        }
    }

    /**
     * Copies known passwords onto users received from broker, since we get only password hash from it, so that
     * ensurePresent does not fail due to differences in the password field.

     * @param existing   Existing users
     * *
     * @param fromConfig Users from input configuration
     * *
     * @return Existing users with known passwords appended
     */
    private fun withKnownPasswords(existing: Map<String, UserData>,
                                   fromConfig: Map<String, UserData>): Map<String, UserData> {
        existing.forEach { entry ->
            entry.value.password = fromConfig.getOrElse(entry.key, {""}).toString()
        }
        return existing
    }

    /**
     * Attempts to find a user in the given configuration with rights to configure the requested resource.

     * @param permissions  user permissions
     * *
     * @param vhost        vhost name
     * *
     * @param resourceName resource name
     * *
     * @return Optional of user with creation rights, or Optional.empty() if not found.
     */
    private fun authForResource(users: Map<String, UserData>,
                                permissions: Map<String, PermissionsData>,
                                vhost: String,
                                resourceName: String): Pair<String, String> {

        val auth = permissions.entries.filter({ entry ->
            // Find permission with rights to edit resourceName on vhost
            var res = false
            val matcher = atVHostPattern.matcher(entry.key)
            if (matcher.matches() && matcher.group(2) == vhost) {
                if (Pattern.compile(entry.value.configure).matcher(resourceName).matches()) {
                    res = true
                }
            }
            res
        }).map({ entry ->
            // Extract user name from permission
            val matcher = atVHostPattern.matcher(entry.key)
            matcher.find() // Causes matcher to actually parse regex groups
            matcher.group(1)
        }).filter({
            // Filter away user if it isn't specified in configuration
            user -> users.contains(user)
        }).map({ user ->
            Pair(user, users[user]!!.password)
        }).firstOrNull()

        if (auth != null) {
            return auth
        } else {
            return defaultAuth()
        }
    }

    private fun defaultAuth(): Pair<String, String> = Pair(client.getUsername(), client.getPassword())
}
